### DISCUSSION TITLE
how to compare or distinguish two different websockets in fastapi?

### DISCUSSION THREAD

**Discussion by DiponRoy** [fastapi/fastapi#6857] [Questions] [answered]
### Front End
From the same page, if Connect button clicked multiple times, multiple WebSocket requests landing in the backend.

```Python
<!DOCTYPE html>
<html>
    <head>
        <title>Chat</title>
    </head>
    <body>
        <h1>WebSocket Chat</h1>
        <form action="" onsubmit="sendMessage(event)">
            <label>Token: <input type="text" id="token" autocomplete="off" value="some-key-token"/></label>
            <button onclick="connect(event)">Connect</button>
            <hr>
            <!-- <label>Message: <input type="text" id="messageText" autocomplete="off"/></label> -->
            <!-- <button>Send</button> -->
        </form>
        <ul id='messages'>
        </ul>
        <script>
        var ws = null;
            function connect(event) {
                var token = document.getElementById("token")
                ws = new WebSocket("ws://localhost:6003/ws/v1/inbox/test?token=" + token.value);
				
				ws.onopen = function () {
				  console.log('socket opened'); 
				};
                ws.onmessage = function(event) {
                    var messages = document.getElementById('messages')
                    var message = document.createElement('li')
					var content = document.createTextNode(event.data)
                    <!-- var data = document.createTextNode(event.data) -->
					<!-- var content = "message:" + data.message -->
                    message.appendChild(content)
                    messages.appendChild(message)
                };
				
				ws.onclose = function(e) {  
				  console.log('socket closed from server'); 
				}

				ws.onerror = function(err) {
				  console.error(err)
				};
				
                event.preventDefault()
            }
            function sendMessage(event) {
                var input = document.getElementById("messageText")
                ws.send(input.value)
                input.value = ''
                event.preventDefault()
            }
        </script>
    </body>
</html>
```

### Back End
Here we are adding incoming WebSockets to a list. Here we want to compare incoming WebSockets considering the request source. The important thing a user should be able to connect from different devices, browsers, or browser tabs. 

```Python
from fastapi import FastAPI, Request, Depends, WebSocket, WebSocketDisconnect, Query

app = FastAPI()

class WebSocketConnectionModel:
    user_id: int
    account_id: int
    connection_date_time_utc: datetime
    socket: WebSocket


class SocketManager:
    def __init__(self):
        self.connections: List[WebSocketConnectionModel] = []

    def len(self):
        return len(self.connections)

    def connect(self, account_id: int, user_id: int, websocket: WebSocket):
        connection = WebSocketConnectionModel()
        connection.account_id = account_id
        connection.user_id = user_id
        connection.socket = websocket
        connection.connection_date_time_utc = datetime.utcnow()
        self.connections.append(connection)
        return connection

    def disconnect(self, item: WebSocketConnectionModel):
        self.connections.remove(item)



manager = SocketManager()

async def websocket_test(websocket: WebSocket, access_token: str = Query(..., alias="token")):
    user = get_current_user_from_query_token(access_token)
    await websocket.accept()
   #add connection to a list
    connection = manager.connect(user.account_id, user.id, websocket)
    try:
        while True:
            data = await websocket.receive_json()
    except WebSocketDisconnect:
        manager.disconnect(connection)

app.add_websocket_route("/ws/v1/inbox/test", facebook_inbox.websocket_test)

```

Is there any way to identify incoming WebSocket object and compare with an existing list, to avoid adding duplicates? Is there any unique identifier in WebSocket object?

**Comment by aalifadv** [comment, marked as answer]
> Is there any way to identify incoming WebSocket object and compare with an existing list to avoid adding duplicates? Is there any unique identifier in WebSocket object?

This really depends on what you consider a unique identifier and what goes into formulating the unique identifier. WebSocket allow the pass of paramaters  in the header, query and path. You can use the passed information from the client to generate unique identifier in the backend as you desire and then store it.  It can be a simple hash function that takes what you think should formulate unique identifiers e.g device_info:browser_info:client_ip -> uniquID.

I have never seen this handled by a framework in the past so I dont think FastAPI would be the right place to add such a logic.

**Comment by stevenj** [comment]
@aalifadv is correct.  IF you have any authentication (i use basic auth over tls) you would use that to uniquely identify a client.  If a client can connect multiple times, and you need to distinguish their connections, generate a UUID on connection in the client and send it in a header, or send it as the very first message from the client.  This is an example of ways you could solve this.  Websockets are very "Laissez-faire" so your not going to find a bunch of prescriptive rules telling you how you must communicate, but with that freedom comes the responsibility to work out issues like this for your self.  If you work out a way that works for you, it is by definition, correct.

**Comment by DiponRoy** [comment]
@aalifadv and @stevenj thank you for the idea, it will work for me.

**Comment by redigaffi** [comment]
@aalifadv Well, in socket.IO (node.js) you actually get an identifier that you can access with: socket.id
